File: packages/cojson/src/coValueCore/SessionMap.ts
import { ControlledAccountOrAgent } from "../coValues/account.js";
import { TRANSACTION_CONFIG } from "../config.js";
import type {
  CryptoProvider,
  KeyID,
  KeySecret,
  SessionMapImpl,
  Signature,
  SignerID,
} from "../crypto/crypto.js";
import { isDeleteSessionID, RawCoID, SessionID } from "../ids.js";
import { parseJSON, stableStringify, Stringified } from "../jsonStringify.js";
import { JsonObject, JsonValue } from "../jsonValue.js";
import { Transaction, CoValueHeader } from "./verifiedState.js";
import {
  CoValueKnownState,
  KnownStateSessions,
  cloneKnownState,
} from "../knownState.js";

export type SessionLog = {
  signerID?: SignerID;
  transactions: Transaction[];
  lastSignature: Signature | undefined;
  signatureAfter: { [txIdx: number]: Signature | undefined };
  sessionID: SessionID;
};

/**
 * SessionMap wraps the Rust SessionMapImpl and provides a TypeScript-friendly interface.
 * All transaction storage and crypto operations are delegated to Rust.
 */
export class SessionMap {
  private readonly impl: SessionMapImpl;
  private readonly id: RawCoID;
  private readonly crypto: CryptoProvider;

  // Cache for SessionLog objects to avoid re-parsing on every access
  private sessionLogCache: Map<SessionID, SessionLog> = new Map();
  private sessionLogCacheValid: Map<SessionID, number> = new Map(); // txCount when cached

  // Cache for immutable known state (same reference until invalidated)
  private cachedImmutableKnownState: CoValueKnownState | undefined;
  private cachedImmutableKnownStateWithStreaming: CoValueKnownState | undefined;

  constructor(
    id: RawCoID,
    crypto: CryptoProvider,
    header: CoValueHeader,
    streamingKnownState?: KnownStateSessions,
  ) {
    this.id = id;
    this.crypto = crypto;

    // Create the Rust SessionMapImpl with the header and max tx size threshold
    const headerJson = stableStringify(header);
    this.impl = crypto.createSessionMap(
      id,
      headerJson,
      TRANSACTION_CONFIG.MAX_RECOMMENDED_TX_SIZE,
    );

    // Set streaming known state if provided
    if (streamingKnownState) {
      this.impl.setStreamingKnownState(JSON.stringify(streamingKnownState));
    }
  }

  // Private constructor for cloning
  private static fromImpl(
    id: RawCoID,
    crypto: CryptoProvider,
    impl: SessionMapImpl,
  ): SessionMap {
    const instance = Object.create(SessionMap.prototype) as SessionMap;
    (instance as any).id = id;
    (instance as any).crypto = crypto;
    (instance as any).impl = impl;
    (instance as any).sessionLogCache = new Map();
    (instance as any).sessionLogCacheValid = new Map();
    return instance;
  }

  markAsDeleted() {
    this.impl.markAsDeleted();
    this.invalidateCache();
  }

  get isDeleted(): boolean {
    return this.impl.isDeleted();
  }

  setStreamingKnownState(streamingKnownState: KnownStateSessions) {
    if (this.isDeleted) {
      return;
    }
    this.impl.setStreamingKnownState(JSON.stringify(streamingKnownState));
  }

  get knownState(): CoValueKnownState {
    // Native object returned directly from Rust
    return this.impl.getKnownState() as CoValueKnownState;
  }

  get knownStateWithStreaming(): CoValueKnownState | undefined {
    // Native object returned directly from Rust
    const result = this.impl.getKnownStateWithStreaming();
    if (!result || result === undefined) return undefined;
    return result as CoValueKnownState;
  }

  getImmutableKnownState(): CoValueKnownState {
    if (!this.cachedImmutableKnownState) {
      this.cachedImmutableKnownState = cloneKnownState(this.knownState);
    }
    return this.cachedImmutableKnownState;
  }

  getImmutableKnownStateWithStreaming(): CoValueKnownState {
    const withStreaming = this.knownStateWithStreaming;
    if (!withStreaming) {
      return this.getImmutableKnownState();
    }
    if (!this.cachedImmutableKnownStateWithStreaming) {
      this.cachedImmutableKnownStateWithStreaming =
        cloneKnownState(withStreaming);
    }
    return this.cachedImmutableKnownStateWithStreaming;
  }

  get(sessionID: SessionID): SessionLog | undefined {
    const txCount = this.impl.getTransactionCount(sessionID);
    if (txCount === -1) {
      return undefined;
    }
    return this.getSessionLog(sessionID);
  }

  getTransactionsCount(sessionID: SessionID): number | undefined {
    const txCount = this.impl.getTransactionCount(sessionID);
    if (txCount === -1) {
      return undefined;
    }
    return txCount;
  }

  private invalidateCache() {
    this.sessionLogCache.clear();
    this.sessionLogCacheValid.clear();
    // Invalidate immutable known state caches
    this.cachedImmutableKnownState = undefined;
    this.cachedImmutableKnownStateWithStreaming = undefined;
  }

  /**
   * Update the session log cache directly when adding transactions.
   * This avoids round-trips to Rust on subsequent reads.
   */
  private updateSessionLogCache(
    sessionID: SessionID,
    signerID: SignerID | undefined,
    newTransactions: Transaction[],
    newSignature: Signature,
  ) {
    const cached = this.sessionLogCache.get(sessionID);
    const currentTxCount = this.impl.getTransactionCount(sessionID);

    if (cached) {
      // Append to existing cache
      cached.transactions.push(...newTransactions);
      cached.lastSignature = newSignature;
      if (signerID) {
        cached.signerID = signerID;
      }
      // Check if we need to update signatureAfter (in-between signature)
      const lastCheckpoint = this.impl.getLastSignatureCheckpoint(sessionID);
      if (
        lastCheckpoint !== undefined &&
        lastCheckpoint !== null &&
        lastCheckpoint >= 0
      ) {
        const sig = this.impl.getSignatureAfter(sessionID, lastCheckpoint);
        if (sig) {
          cached.signatureAfter[lastCheckpoint] = sig as Signature;
        }
      }
      this.sessionLogCacheValid.set(sessionID, currentTxCount);
    } else {
      // Create new cache entry
      const sessionLog: SessionLog = {
        signerID,
        transactions: [...newTransactions],
        lastSignature: newSignature,
        signatureAfter: {},
        sessionID,
      };
      // Check for in-between signature
      const lastCheckpoint = this.impl.getLastSignatureCheckpoint(sessionID);
      if (
        lastCheckpoint !== undefined &&
        lastCheckpoint !== null &&
        lastCheckpoint >= 0
      ) {
        const sig = this.impl.getSignatureAfter(sessionID, lastCheckpoint);
        if (sig) {
          sessionLog.signatureAfter[lastCheckpoint] = sig as Signature;
        }
      }
      this.sessionLogCache.set(sessionID, sessionLog);
      this.sessionLogCacheValid.set(sessionID, currentTxCount);
    }
  }

  private getSessionLog(sessionID: SessionID): SessionLog {
    const currentTxCount = this.impl.getTransactionCount(sessionID);
    const cachedTxCount = this.sessionLogCacheValid.get(sessionID);

    // Check if cache is valid
    if (cachedTxCount === currentTxCount) {
      const cached = this.sessionLogCache.get(sessionID);
      if (cached) return cached;
    }

    // Build fresh SessionLog from Rust data
    const transactions: Transaction[] = [];
    const signatureAfter: { [txIdx: number]: Signature | undefined } = {};

    // Fetch all transactions
    if (currentTxCount > 0) {
      // getSessionTransactions returns array of JSON strings directly
      const txStrings = this.impl.getSessionTransactions(sessionID, 0);
      if (txStrings) {
        for (const txStr of txStrings) {
          transactions.push(JSON.parse(txStr) as Transaction);
        }
      }
    }

    // Build signatureAfter map
    const lastCheckpoint = this.impl.getLastSignatureCheckpoint(sessionID);
    if (
      lastCheckpoint !== undefined &&
      lastCheckpoint !== null &&
      lastCheckpoint >= 0
    ) {
      // We need to find all checkpoints - iterate from 0 to lastCheckpoint
      for (let i = 0; i <= lastCheckpoint; i++) {
        const sig = this.impl.getSignatureAfter(sessionID, i);
        if (sig) {
          signatureAfter[i] = sig as Signature;
        }
      }
    }

    const lastSignature = this.impl.getLastSignature(sessionID) as
      | Signature
      | undefined;

    const sessionLog: SessionLog = {
      signerID: undefined, // We don't track this in Rust currently
      transactions,
      lastSignature,
      signatureAfter,
      sessionID,
    };

    // Cache the result
    this.sessionLogCache.set(sessionID, sessionLog);
    this.sessionLogCacheValid.set(sessionID, currentTxCount);

    return sessionLog;
  }

  addTransaction(
    sessionID: SessionID,
    signerID: SignerID | undefined,
    newTransactions: Transaction[],
    newSignature: Signature,
    skipVerify: boolean = false,
  ) {
    if (this.isDeleted && !isDeleteSessionID(sessionID)) {
      throw new Error("Cannot add transactions to a deleted coValue");
    }

    // Convert transactions to JSON array
    const txJson = JSON.stringify(newTransactions);

    this.impl.addTransactions(
      sessionID,
      signerID,
      txJson,
      newSignature,
      skipVerify,
    );

    // Update cache directly instead of invalidating
    this.updateSessionLogCache(
      sessionID,
      signerID,
      newTransactions,
      newSignature,
    );
    // Invalidate immutable known state caches
    this.cachedImmutableKnownState = undefined;
    this.cachedImmutableKnownStateWithStreaming = undefined;
  }

  makeNewPrivateTransaction(
    sessionID: SessionID,
    signerAgent: ControlledAccountOrAgent,
    changes: JsonValue[],
    keyID: KeyID,
    keySecret: KeySecret,
    meta: JsonObject | undefined,
    madeAt: number,
  ): { signature: Signature; transaction: Transaction } {
    if (this.isDeleted) {
      throw new Error(
        "Cannot make new private transaction on a deleted coValue",
      );
    }

    const changesJson = JSON.stringify(changes);
    const metaJson = meta ? JSON.stringify(meta) : undefined;
    const signerSecret = signerAgent.currentSignerSecret();

    const resultJson = this.impl.makeNewPrivateTransaction(
      sessionID,
      signerSecret,
      changesJson,
      keyID,
      keySecret,
      metaJson,
      madeAt,
    );

    const result = JSON.parse(resultJson) as {
      signature: string;
      transaction: Transaction;
    };

    const signature = result.signature as Signature;

    // Update cache directly instead of invalidating
    this.updateSessionLogCache(
      sessionID,
      signerAgent.id as unknown as SignerID,
      [result.transaction],
      signature,
    );
    // Invalidate immutable known state caches
    this.cachedImmutableKnownState = undefined;
    this.cachedImmutableKnownStateWithStreaming = undefined;

    return {
      signature,
      transaction: result.transaction,
    };
  }

  makeNewTrustingTransaction(
    sessionID: SessionID,
    signerAgent: ControlledAccountOrAgent,
    changes: JsonValue[],
    meta: JsonObject | undefined,
    madeAt: number,
  ): { signature: Signature; transaction: Transaction } {
    if (this.isDeleted) {
      throw new Error(
        "Cannot make new trusting transaction on a deleted coValue",
      );
    }

    const changesJson = JSON.stringify(changes);
    const metaJson = meta ? JSON.stringify(meta) : undefined;
    const signerSecret = signerAgent.currentSignerSecret();

    const resultJson = this.impl.makeNewTrustingTransaction(
      sessionID,
      signerSecret,
      changesJson,
      metaJson,
      madeAt,
    );

    const result = JSON.parse(resultJson) as {
      signature: string;
      transaction: Transaction;
    };

    const signature = result.signature as Signature;

    // Update cache directly instead of invalidating
    this.updateSessionLogCache(
      sessionID,
      signerAgent.id as unknown as SignerID,
      [result.transaction],
      signature,
    );
    // Invalidate immutable known state caches
    this.cachedImmutableKnownState = undefined;
    this.cachedImmutableKnownStateWithStreaming = undefined;

    return {
      signature,
      transaction: result.transaction,
    };
  }

  decryptTransaction(
    sessionID: SessionID,
    txIndex: number,
    keySecret: KeySecret,
  ): JsonValue[] | undefined {
    const decrypted = this.impl.decryptTransaction(
      sessionID,
      txIndex,
      keySecret,
    );
    if (!decrypted) {
      return undefined;
    }
    return parseJSON(decrypted as Stringified<JsonValue[]>);
  }

  decryptTransactionMeta(
    sessionID: SessionID,
    txIndex: number,
    keySecret: KeySecret,
  ): JsonObject | undefined {
    const sessionLog = this.get(sessionID);
    if (!sessionLog?.transactions[txIndex]?.meta) {
      return undefined;
    }
    const decrypted = this.impl.decryptTransactionMeta(
      sessionID,
      txIndex,
      keySecret,
    );
    if (!decrypted) {
      return undefined;
    }
    return parseJSON(decrypted as Stringified<JsonObject>);
  }

  get size(): number {
    const sessionIds = JSON.parse(this.impl.getSessionIds()) as string[];
    return sessionIds.length;
  }

  get sessions(): Map<SessionID, SessionLog> {
    // Build a Map from all sessions
    const map = new Map<SessionID, SessionLog>();
    const sessionIds = JSON.parse(this.impl.getSessionIds()) as SessionID[];
    for (const sessionID of sessionIds) {
      map.set(sessionID, this.getSessionLog(sessionID));
    }
    return map;
  }

  *entries(): IterableIterator<[SessionID, SessionLog]> {
    const sessionIds = JSON.parse(this.impl.getSessionIds()) as SessionID[];
    for (const sessionID of sessionIds) {
      yield [sessionID, this.getSessionLog(sessionID)];
    }
  }

  *values(): IterableIterator<SessionLog> {
    const sessionIds = JSON.parse(this.impl.getSessionIds()) as SessionID[];
    for (const sessionID of sessionIds) {
      yield this.getSessionLog(sessionID);
    }
  }

  *keys(): IterableIterator<SessionID> {
    const sessionIds = JSON.parse(this.impl.getSessionIds()) as SessionID[];
    for (const sessionID of sessionIds) {
      yield sessionID;
    }
  }

  clone(): SessionMap {
    // For cloning, we need to create a new SessionMap with the same state
    // Since Rust SessionMapImpl doesn't have a clone method exposed,
    // we need to recreate the state

    // Get header from Rust
    const headerJson = this.impl.getHeader();
    const header = JSON.parse(headerJson) as CoValueHeader;

    // Get streaming known state
    const knownStateWithStreaming = this.knownStateWithStreaming;
    let streamingKnownState: KnownStateSessions | undefined;
    if (knownStateWithStreaming) {
      // Calculate streaming state as difference between withStreaming and base
      const baseKnownState = this.knownState;
      streamingKnownState = {};
      for (const [sessionId, count] of Object.entries(
        knownStateWithStreaming.sessions,
      )) {
        const baseCount = baseKnownState.sessions[sessionId as SessionID] ?? 0;
        if (count > baseCount) {
          streamingKnownState[sessionId as SessionID] = count;
        }
      }
    }

    // Create new SessionMap
    const clone = new SessionMap(
      this.id,
      this.crypto,
      header,
      streamingKnownState,
    );

    // Copy all sessions with their transactions
    const sessionIds = JSON.parse(this.impl.getSessionIds()) as SessionID[];
    for (const sessionID of sessionIds) {
      const txCount = this.impl.getTransactionCount(sessionID);
      if (txCount > 0) {
        // getSessionTransactions returns string[] directly
        const txStrings = this.impl.getSessionTransactions(sessionID, 0);
        if (txStrings) {
          const transactions = txStrings.map(
            (s) => JSON.parse(s) as Transaction,
          );
          const lastSignature = this.impl.getLastSignature(sessionID) as
            | Signature
            | undefined;
          if (lastSignature) {
            clone.impl.addTransactions(
              sessionID,
              undefined, // signerID not tracked
              JSON.stringify(transactions),
              lastSignature,
              true, // skip verify since we're cloning valid data
            );
          }
        }
      }
    }

    // Copy deletion state
    if (this.isDeleted) {
      clone.markAsDeleted();
    }

    return clone;
  }
}

File: crates/cojson-core-rn/rust/src/session_map.rs
use cojson_core::core::{KnownState as RustKnownState, SessionMapImpl};
use std::collections::HashMap;
use thiserror::Error;

#[derive(Error, Debug, uniffi::Error)]
pub enum SessionMapError {
    #[error("SessionMap error: {0}")]
    Internal(String),
    #[error("Failed to acquire lock")]
    LockError,
}

/// KnownState as a native Record (no JSON serialization needed)
#[derive(uniffi::Record, Clone, Debug)]
pub struct KnownState {
    pub id: String,
    pub header: bool,
    pub sessions: HashMap<String, u32>,
}

impl From<&RustKnownState> for KnownState {
    fn from(ks: &RustKnownState) -> Self {
        KnownState {
            id: ks.id.clone(),
            header: ks.header,
            sessions: ks.sessions.iter().map(|(k, v)| (k.clone(), *v)).collect(),
        }
    }
}

#[derive(uniffi::Object)]
pub struct SessionMap {
    internal: std::sync::Mutex<SessionMapImpl>,
}

#[uniffi::export]
impl SessionMap {
    /// Create a new SessionMap for a CoValue
    /// Create a new SessionMap for a CoValue.
    /// Validates the header and verifies that `co_id` matches the hash of the header.
    /// `max_tx_size` is the threshold for recording in-between signatures (default: 100KB)
    /// `skip_verify` if true, skips uniqueness and ID validation (for trusted storage shards)
    #[uniffi::constructor]
    pub fn new(
        co_id: String,
        header_json: String,
        max_tx_size: Option<u32>,
        skip_verify: Option<bool>,
    ) -> Result<Self, SessionMapError> {
        let internal = SessionMapImpl::new_with_skip_verify(
            &co_id,
            &header_json,
            max_tx_size,
            skip_verify.unwrap_or(false),
        )
        .map_err(|e| SessionMapError::Internal(e.to_string()))?;
        Ok(SessionMap {
            internal: std::sync::Mutex::new(internal),
        })
    }

    // === Header ===

    /// Get the header as JSON
    pub fn get_header(&self) -> Result<String, SessionMapError> {
        let internal = self
            .internal
            .lock()
            .map_err(|_| SessionMapError::LockError)?;
        Ok(internal.get_header())
    }

    // === Transaction Operations ===

    /// Add transactions to a session
    pub fn add_transactions(
        &self,
        session_id: String,
        signer_id: Option<String>,
        transactions_json: String,
        signature: String,
        skip_verify: bool,
    ) -> Result<(), SessionMapError> {
        let mut internal = self
            .internal
            .lock()
            .map_err(|_| SessionMapError::LockError)?;
        internal
            .add_transactions(
                &session_id,
                signer_id.as_deref(),
                &transactions_json,
                &signature,
                skip_verify,
            )
            .map_err(|e| SessionMapError::Internal(e.to_string()))
    }

    /// Create new private transaction (for local writes)
    /// Returns JSON: { signature: string, transaction: Transaction }
    pub fn make_new_private_transaction(
        &self,
        session_id: String,
        signer_secret: String,
        changes_json: String,
        key_id: String,
        key_secret: String,
        meta_json: Option<String>,
        made_at: f64,
    ) -> Result<String, SessionMapError> {
        let mut internal = self
            .internal
            .lock()
            .map_err(|_| SessionMapError::LockError)?;
        let signed_tx = internal
            .make_new_private_transaction(
                session_id,
                signer_secret,
                &changes_json,
                key_id,
                key_secret,
                meta_json,
                made_at as u64,
            )
            .map_err(|e| SessionMapError::Internal(e.to_string()))?;

        let tx_json = serde_json::to_string(&signed_tx.transaction)
            .map_err(|e| SessionMapError::Internal(e.to_string()))?;
        Ok(format!(
            r#"{{"signature":"{}","transaction":{}}}"#,
            signed_tx.signature.0, tx_json
        ))
    }

    /// Create new trusting transaction (for local writes)
    /// Returns JSON: { signature: string, transaction: Transaction }
    pub fn make_new_trusting_transaction(
        &self,
        session_id: String,
        signer_secret: String,
        changes_json: String,
        meta_json: Option<String>,
        made_at: f64,
    ) -> Result<String, SessionMapError> {
        let mut internal = self
            .internal
            .lock()
            .map_err(|_| SessionMapError::LockError)?;
        let signed_tx = internal
            .make_new_trusting_transaction(
                session_id,
                signer_secret,
                &changes_json,
                meta_json,
                made_at as u64,
            )
            .map_err(|e| SessionMapError::Internal(e.to_string()))?;

        let tx_json = serde_json::to_string(&signed_tx.transaction)
            .map_err(|e| SessionMapError::Internal(e.to_string()))?;
        Ok(format!(
            r#"{{"signature":"{}","transaction":{}}}"#,
            signed_tx.signature.0, tx_json
        ))
    }

    // === Session Queries ===

    /// Get all session IDs as native array
    pub fn get_session_ids(&self) -> Result<Vec<String>, SessionMapError> {
        let internal = self
            .internal
            .lock()
            .map_err(|_| SessionMapError::LockError)?;
        Ok(internal.get_session_ids())
    }

    /// Get transaction count for a session (returns -1 if session not found)
    pub fn get_transaction_count(&self, session_id: String) -> Result<i32, SessionMapError> {
        let internal = self
            .internal
            .lock()
            .map_err(|_| SessionMapError::LockError)?;
        Ok(internal
            .get_transaction_count(&session_id)
            .map(|c| c as i32)
            .unwrap_or(-1))
    }

    /// Get single transaction by index as JSON string (returns None if not found)
    pub fn get_transaction(
        &self,
        session_id: String,
        tx_index: u32,
    ) -> Result<Option<String>, SessionMapError> {
        let internal = self
            .internal
            .lock()
            .map_err(|_| SessionMapError::LockError)?;
        Ok(internal.get_transaction(&session_id, tx_index))
    }

    /// Get transactions for a session from index as JSON strings (returns None if session not found)
    pub fn get_session_transactions(
        &self,
        session_id: String,
        from_index: u32,
    ) -> Result<Option<Vec<String>>, SessionMapError> {
        let internal = self
            .internal
            .lock()
            .map_err(|_| SessionMapError::LockError)?;
        Ok(internal.get_session_transactions(&session_id, from_index))
    }

    /// Get last signature for a session (returns None if session not found)
    pub fn get_last_signature(
        &self,
        session_id: String,
    ) -> Result<Option<String>, SessionMapError> {
        let internal = self
            .internal
            .lock()
            .map_err(|_| SessionMapError::LockError)?;
        Ok(internal.get_last_signature(&session_id))
    }

    /// Get signature after specific transaction index
    pub fn get_signature_after(
        &self,
        session_id: String,
        tx_index: u32,
    ) -> Result<Option<String>, SessionMapError> {
        let internal = self
            .internal
            .lock()
            .map_err(|_| SessionMapError::LockError)?;
        Ok(internal.get_signature_after(&session_id, tx_index))
    }

    /// Get the last signature checkpoint index (-1 if no checkpoints, None if session not found)
    pub fn get_last_signature_checkpoint(
        &self,
        session_id: String,
    ) -> Result<Option<i32>, SessionMapError> {
        let internal = self
            .internal
            .lock()
            .map_err(|_| SessionMapError::LockError)?;
        Ok(internal.get_last_signature_checkpoint(&session_id))
    }

    // === Known State ===

    /// Get the known state as a native Record
    pub fn get_known_state(&self) -> Result<KnownState, SessionMapError> {
        let internal = self
            .internal
            .lock()
            .map_err(|_| SessionMapError::LockError)?;
        Ok(internal.get_known_state().into())
    }

    /// Get the known state with streaming as a native Record
    pub fn get_known_state_with_streaming(&self) -> Result<Option<KnownState>, SessionMapError> {
        let internal = self
            .internal
            .lock()
            .map_err(|_| SessionMapError::LockError)?;
        Ok(internal
            .get_known_state_with_streaming()
            .map(|ks| ks.into()))
    }

    /// Set streaming known state
    pub fn set_streaming_known_state(&self, streaming_json: String) -> Result<(), SessionMapError> {
        let mut internal = self
            .internal
            .lock()
            .map_err(|_| SessionMapError::LockError)?;
        internal
            .set_streaming_known_state(&streaming_json)
            .map_err(|e| SessionMapError::Internal(e.to_string()))
    }

    // === Deletion ===

    /// Mark this CoValue as deleted
    pub fn mark_as_deleted(&self) -> Result<(), SessionMapError> {
        let mut internal = self
            .internal
            .lock()
            .map_err(|_| SessionMapError::LockError)?;
        internal.mark_as_deleted();
        Ok(())
    }

    /// Check if this CoValue is deleted
    pub fn is_deleted(&self) -> Result<bool, SessionMapError> {
        let internal = self
            .internal
            .lock()
            .map_err(|_| SessionMapError::LockError)?;
        Ok(internal.is_deleted())
    }

    // === Decryption ===

    /// Decrypt transaction changes
    pub fn decrypt_transaction(
        &self,
        session_id: String,
        tx_index: u32,
        key_secret: String,
    ) -> Result<Option<String>, SessionMapError> {
        let internal = self
            .internal
            .lock()
            .map_err(|_| SessionMapError::LockError)?;
        internal
            .decrypt_transaction(&session_id, tx_index, &key_secret)
            .map_err(|e| SessionMapError::Internal(e.to_string()))
    }

    /// Decrypt transaction meta
    pub fn decrypt_transaction_meta(
        &self,
        session_id: String,
        tx_index: u32,
        key_secret: String,
    ) -> Result<Option<String>, SessionMapError> {
        let internal = self
            .internal
            .lock()
            .map_err(|_| SessionMapError::LockError)?;
        internal
            .decrypt_transaction_meta(&session_id, tx_index, &key_secret)
            .map_err(|e| SessionMapError::Internal(e.to_string()))
    }
}
